File: IOSViperSample/IOSViperSample/chapterDetail/ChapterDetailsPresenterImpl.swift
//
//  ChapterDetailsPresenterImpl.swift
//  IOSViperSample
//
//  Created by Ivan Phytsyk on 10/11/18.
//  Copyright © 2018 Ivan Phytsyk. All rights reserved.
//

import Foundation

class ChapterDetailsPresenterImpl: ChapterDetailsPresenter {
    private let chapterDetailsInteractor: ChapterDetailsInteractor
    private let view: ChapterDetailsView
    private let router: ChapterDetailsRouter
    private let chapterTitle: String
    
    init(_ chapterTitle: String,
         _ chapterDetailsInteractor: ChapterDetailsInteractor
        ,_ view: ChapterDetailsView
        ,_ router: ChapterDetailsRouter) {
        self.chapterTitle = chapterTitle
        self.chapterDetailsInteractor = chapterDetailsInteractor
        self.view = view
        self.router = router
    }
    
    func start() {
        chapterDetailsInteractor.execute(input: self.chapterTitle, {chapter in
            let viewModel = ChapterViewModel((chapter?.title)!, (chapter?.description)!, getIconForChapterType((chapter?.type)!))
            self.view.showChapterDetails(viewModel)
        }, {error in
            self.view.showErrorMessage(error.localizedDescription)
        }
        )
    
    }
    
    func stop() {
        chapterDetailsInteractor.cancel()
    }
    
    
}

File: AndroidViperSample/app/src/main/java/fytsyk/com/vipersample/chapterDetails/ChapterDetailsPresenterImpl.kt
package fytsyk.com.vipersample.chapterDetails

import fytsyk.com.vipersample.chapterDetails.viper.ChapterDetailsInteractor
import fytsyk.com.vipersample.chapterDetails.viper.ChapterDetailsPresenter
import fytsyk.com.vipersample.chapterDetails.viper.ChapterDetailsRouter
import fytsyk.com.vipersample.chapterDetails.viper.ChapterDetailsView
import fytsyk.com.vipersample.chapters.viper.ChapterViewModel
import fytsyk.com.vipersample.chapters.viper.getIconForChapterType

class ChapterDetailsPresenterImpl(private val chapterId: String,
                                  private val chapterDetailsInteractor: ChapterDetailsInteractor,
                                  private val view: ChapterDetailsView,
                                  private val router: ChapterDetailsRouter) : ChapterDetailsPresenter {
    override fun start() {
        chapterDetailsInteractor.execute(chapterId, {
            val viewModel = ChapterViewModel(it?.title!!, it.description, getIconForChapterType(it.type))
            view.showChapterDetails(viewModel)
        }, {
            view.showErrorMessage(it.localizedMessage)
        })
    }

    override fun stop() {
        chapterDetailsInteractor.cancel()
    }
}